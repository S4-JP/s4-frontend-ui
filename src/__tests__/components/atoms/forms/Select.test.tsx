import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import { Select, Color, Icon } from '../../../../index'

/**
 * type Props = {
 *   name?: string
 *   defaultData: {
 *     value: string
 *     label: string
 *     image?: string
 *     icon?: React.ReactNode
 *   }
 *   data?: any[]
 *   top?: boolean
 *   className?: string
 *   disabled?: boolean
 *   onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void
 *   handleClick?(e: React.MouseEvent<HTMLElement>): void
 * }
 */

type DataType = {
  value: string
  label: string
  image?: string
  icon?: React.ReactNode
}
const defaultData: DataType = {
  value: 'test value',
  label: 'test label',
}
const testData: DataType[] = [
  { value: 'test value', label: 'test label' },
  { value: 'test value 1', label: 'test label 1' },
]
const testDataWithIcon: DataType[] = [
  {
    value: 'test value with image',
    label: 'test label with image',
    image: 'test_src',
  },
  {
    value: 'test value with icon',
    label: 'test label with icon',
    icon: <Icon.Cation color={Color.TEXT.GRAY} />,
  },
]

const mockHandleClick = jest.fn()

describe('form - Select のレンダリング確認', () => {
  test('オプションなしで設定した場合、初期値でレンダリングされる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testData}
        handleClick={mockHandleClick}
      />
    )

    expect(getByTestId('atoms-f-select-wrap')).toMatchSnapshot()
    expect(screen.getByTestId('atoms-f-select-options')).toHaveStyle(
      'top: 50px;'
    )
  })

  test('disabledを設定した場合、selectがdisableになる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testData}
        handleClick={mockHandleClick}
        disabled
      />
    )

    expect(getByTestId('atoms-f-select-wrap')).toHaveAttribute('disabled', '')
  })

  test('SelectItemをクリックした場合、props.handleClickがコールされる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testData}
        handleClick={mockHandleClick}
        disabled
      />
    )

    fireEvent.click(getByTestId('atoms-f-select-options-input-0'))
    expect(mockHandleClick).toBeCalledTimes(1)
  })

  test('SelectItemをクリックした場合、SelectとOptionsがフォーカスされる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testData}
        handleClick={mockHandleClick}
        disabled
      />
    )

    fireEvent.click(getByTestId('atoms-f-select-selected-item'))
    waitFor(() =>
      expect(getByTestId('atoms-f-select-selected-item')).toHaveFocus()
    )
    waitFor(() => expect(getByTestId('atoms-f-select-options')).toHaveFocus())
  })

  test('SelectItemにimageとiconをセットした場合、レンダリングされる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testDataWithIcon}
        handleClick={mockHandleClick}
        disabled
      />
    )

    expect(getByTestId('atoms-f-select-options-input-0')).toMatchSnapshot()
    expect(getByTestId('atoms-f-select-options-input-1')).toMatchSnapshot()
  })

  test('props.topを設定した場合、設定値でレンダリングされる。', () => {
    const { getByTestId } = render(
      <Select
        defaultData={defaultData}
        data={testDataWithIcon}
        handleClick={mockHandleClick}
        top={true}
      />
    )
    expect(getByTestId('atoms-f-select-options')).toHaveStyle('bottom: 50px;')
  })
})
