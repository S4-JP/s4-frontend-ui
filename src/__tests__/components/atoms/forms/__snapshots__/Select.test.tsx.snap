// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form - Select のレンダリング確認 SelectItemにimageとiconをセットした場合、レンダリングされる。 1`] = `
<input
  data-testid="atoms-f-select-options-input-0"
  type="radio"
  value="test value with image"
/>
`;

exports[`form - Select のレンダリング確認 SelectItemにimageとiconをセットした場合、レンダリングされる。 2`] = `
<input
  data-testid="atoms-f-select-options-input-1"
  type="radio"
  value="test value with icon"
/>
`;

exports[`form - Select のレンダリング確認 オプションなしで設定した場合、初期値でレンダリングされる。 1`] = `
<div
  class="sc-jIZahH cQBDCM"
  data-testid="atoms-f-select-wrap"
>
  <div
    class="sc-himrzO dYECtE"
    data-testid="atoms-f-select-selected-item"
  >
    <div
      class="sc-gXmSlM icBQmI"
      data-testid="atoms-f-select-selected-item-label"
    >
      test label
    </div>
    <div
      class="sc-cCsOjp gcvNLB"
      data-testid="atoms-f-select-selected-item-icon-wrap"
    >
      <span
        class="sc-papXJ kxFMiI"
        color="#222E40"
        data-testid="atoms-i-icon-wrap"
        size="12"
      >
        <svg
          data-testid="atoms-icon-svg-chevron-down"
          viewBox="0 0 40 40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 40V0h40v40z"
            fill="none"
          />
          <path
            d="M36.259 13.349a2.954 2.954 0 0 0-4.086 0L19.999 25.521 7.824 13.346a2.89 2.89 0 0 0-4.086 4.088l14.218 14.218a2.885 2.885 0 0 0 4.086 0l14.217-14.218a2.885 2.885 0 0 0 0-4.086Z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="sc-ciZhAO bilKRs"
    data-testid="atoms-f-select-options"
  >
    <label
      data-testid="atoms-f-select-options-label"
    >
      <label
        class="sc-jdAMXn eICFSo"
      >
        <input
          data-testid="atoms-f-select-options-input-0"
          type="radio"
          value="test value"
        />
        <div
          class="sc-iAvgwm bcoBjA"
          data-testid="atoms-f-select-options-text-wrap-0"
        >
          test label
        </div>
      </label>
      <label
        class="sc-jdAMXn jknONv"
      >
        <input
          data-testid="atoms-f-select-options-input-1"
          type="radio"
          value="test value 1"
        />
        <div
          class="sc-iAvgwm bcoBjA"
          data-testid="atoms-f-select-options-text-wrap-1"
        >
          test label 1
        </div>
      </label>
    </label>
  </div>
</div>
`;
